### PR TITLE
notebooks: factor out data filter logic

### PR DESCRIPTION
**Pull request by rvantonder**
Stacked on https://github.com/sourcegraph/sourcegraph/pull/32636.

This just factors out a record so we can persist those options as-is:

```elm 
type alias DataFilter =
    { dataPoints : Int
    , sortByCount : Bool
    , reverse : Bool
    , excludeStopWords : Bool
    }
```

They correspond to these inputs in the UI:

Screen Shot 2022-03-15 at 6 48 19 PM

These are super experimental options and tied to the current demo. The utility right now outweighs hashing out a more complete way to represent data processing for demo purposes.

After this PR I need just one more that actually persists this, then Sourcegraph users can try it out.

Two commits, to kind of demonstrate how the logic and data decomposes (mostly useful if you have a cursory idea of Elm, if not, don't worry, next week I'll do a walkthrough/workshop, probably emulating exactly this kind of change):

- First commit updates the data type and the `update` function to update that data type
- Second shows how a `view` function can split from the main `view` to only care about the inputs that update the data filter.

## Test plan
Semantics-preserving.

### CODE REVIEW

File: client/web/src/notebooks/blocks/compute/component/src/Main.elm [hunk 1]
```diff
@@ -287,6 +271,31 @@ update msg model =
             ( model, Cmd.none )
 
 
+updateDataFilter : DataFilterMsg -> DataFilter -> DataFilter
+updateDataFilter msg dataFilter =
```
Comment (rvantonder) [inline client/web/src/notebooks/blocks/compute/component/src/Main.elm:275]: @camdencheek here's a concrete example showing that factors out the general `update` case statement to a smaller `updateDataFilter` for just the the substate of the entire model. If this "subcomponent" got larger, you'd split this type and `update` function into its own file and just call it `update`, not `updateDataFilter` (and just compose `Cmd` at the top level if needed)

File: client/web/src/notebooks/blocks/compute/component/src/Main.elm [hunk 2]
```diff
@@ -378,38 +414,7 @@ inputRow model =
...
-                    , label = I.labelRight [] (E.text "reverse")
-                    }
-                , I.checkbox [ E.paddingXY 10 0 ]
-                    { onChange = OnDataFilter << OnExcludeStopWordsCheckbox
-                    , icon = I.defaultCheckbox
-                    , checked = model.dataFilter.excludeStopWords
-                    , label = I.labelRight [] (E.text "exclude stop words")
-                    }
-                ]
+            , E.map OnDataFilter (viewDataFilter model.dataFilter)
```
Comment (rvantonder) [inline client/web/src/notebooks/blocks/compute/component/src/Main.elm:417]: @camdencheek second commit: we can factor out the data filter row to its own view, which returns only the `OnXYZ` data filter messages, that we then wrap with `OnDataFilter` here. Again if this grew larger, you'd just put the factored out view in it's own file, and call it `view`, not `viewDataFilter`

**Review by novoselrok — APPROVED** [review]

File: client/web/src/notebooks/blocks/compute/component/src/Main.elm (hunk 1 above)
Comment (camdencheek) [reply client/web/src/notebooks/blocks/compute/component/src/Main.elm:275]: Niccee. Cool, this makes a lot of sense. This is what I was picturing in my head, but seeing it written down makes it click.

File: client/web/src/notebooks/blocks/compute/component/src/Main.elm [hunk 3]
```diff
@@ -359,6 +368,42 @@ dataView data =
...
+            { onChange = OnDataPoints
+            , placeholder = Nothing
+            , text =
+                case dataFilter.dataPoints of
+                    0 ->
+                        ""
+
+                    n ->
+                        String.fromInt n
+            , label = I.labelHidden ""
```
Comment (camdencheek) [inline client/web/src/notebooks/blocks/compute/component/src/Main.elm:384]: empty label?

**Review by camdencheek — APPROVED** [review]

File: client/web/src/notebooks/blocks/compute/component/src/Main.elm (hunk 3 above)
Comment (rvantonder) [reply client/web/src/notebooks/blocks/compute/component/src/Main.elm:384]: ah, this is specific to the UI element that requires a label for accessibility options: https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element-Input#text. This just suppresses it until I figure out what the Right Thing ™️  to do is.

**Pull request merged**